### ISSUE TITLE
Pydantic secret/payment types are widened to `Any` in lax constructors

### ISSUE THREAD

**Issue by ChenMoFeiJin** [astral-sh/ty#3947]
### Summary

In `ty 0.0.57`, Pydantic types such as `SecretStr`, `SecretBytes`, and `PaymentCardNumber` are widened to `Any` in lax model constructors.

This allows invalid inputs such as `int` and `None` to pass static checking, even though Pydantic rejects them at runtime.

Related to astral-sh/ruff#26587 and astral-sh/ty#2403.

### Reproducer

```python
from pydantic import BaseModel, PaymentCardNumber, SecretBytes, SecretStr


class M(BaseModel):
    password: SecretStr
    token: SecretBytes
    card: PaymentCardNumber


M(password="secret", token=b"token", card="4111111111111111")  # runtime OK

M(password=123, token=b"token", card="4111111111111111")       # runtime error, ty accepts
M(password=None, token=b"token", card="4111111111111111")      # runtime error, ty accepts

M(password="secret", token=123, card="4111111111111111")       # runtime error, ty accepts
M(password="secret", token=None, card="4111111111111111")      # runtime error, ty accepts

M(password="secret", token=b"token", card=123)                 # runtime error, ty accepts
M(password="secret", token=b"token", card=None)                # runtime error, ty accepts
```

### Actual behavior

`ty 0.0.57` infers approximately:

```text
M.__init__:
(self: M, *, password: Any, token: Any, card: Any, **extra: Any) -> None
```

This is too wide.

### Expected behavior

These documented Pydantic types should not fall back to `Any`.

The lax constructor parameter types should be closer to:

```text
SecretStr: str | SecretStr
SecretBytes: bytes | SecretBytes
PaymentCardNumber: str | PaymentCardNumber
```

The exact primitive set can follow Pydantic runtime behavior. The regression is the fallback to `Any`.

### Version

ty 0.0.57

**Comment by sharkdp** [comment]
> Pydantic types such as `SecretStr`, `SecretBytes`, and `PaymentCardNumber` are widened to `Any` in lax model constructors.

Yes. In general, do not expect a type checker to catch all possible mistakes for lax mode models. Types that are not part of the https://pydantic.dev/docs/validation/latest/concepts/conversion_table/ will be mapped to `Any` in lax mode. Even for custom types inside Pydantic, like `SecretStr` or `PaymentCardNumber`, as far as I can tell, there's no way for a type checker to extract the possible input types from those classes (but maybe @Viicos can tell more here?). We'll certainly not maintain a hardcoded huge list of mappings for all possible Pydantic types.

If there is a useful heuristic that we could use (which does not lead to false positives!), I'm open to adding it. But even something like "union of lax-converted constructor input type and class type", which would map `SecretStr` to `LaxStr | SecretStr` is probably not going to work for all types.

**Comment by sharkdp** [comment]
I am tentatively closing this for now, but let me know if someone has ideas on how to improve this. I don't see it as a priority though. Support for strict type checking on lax-mode models will probably always be limited.

**Comment by rodrigoscc** [comment]
Custom Pydantic models are also expanded to Any in lax model constructors. Is this intended? As far as I know, even in lax mode, Pydantic expects an actual instance of the model. In my opinion, Ty should also expect an instance of the specific model, not Any.

Or could we please have a Ty configuration option to disable this new lax type checking? This change has kind of ruined Ty for me :(.

**Comment by sharkdp** [comment]
> Custom Pydantic models are also expanded to Any in lax model constructors. Is this intended?

Yes. I understand that this looks surprising, but this is what Pyrefly and mypy (with the Pydantic plugin enabled) also do. 

> In my opinion, Ty should also expect an instance of the specific model, not Any.

Using the type of the model itself would be too strict, as there are many other things that Pydantic constructors can accept for fields that refer to another model. For example, you can pass in a dictionary:

```py
class Address(BaseModel):
    street: str

class Person(BaseModel):
    name: str
    address: Address

Person(name="Alice", address={"street": "..."})  # accepted at runtime
```

So we currently choose to error on the side of false negatives.

> Or could we please have a Ty configuration option to disable this new lax type checking?

You can already opt out on a per-model (`model_config = ConfigDict(strict=True)`, or `class Person(BaseModel, strict=True`) or even per-field basis (`address: Address = Field(strict=True)`). Does that work for your cases? Obviously, that also changes runtime behavior, but it wasn't clear to me if there's a strong use case for "strict checking at type-check time but lax mode behavior at runtime".

> This change has kind of ruined Ty for me :(.

Sorry to hear that. Can you please describe this in more detail / with an example?

**Comment by rodrigoscc** [comment]
Please take a look at these model definitions:

```python
class Stranger(BaseModel):
    v: int

class Child(BaseModel):
    v: int

class Parent(BaseModel):
    c1: Child
```

Given these model definitions, I see the following error when I pass an instance of `Stranger` to `Parent.c1`:

```
Input should be a valid dictionary or instance of Child [type=model_type, input_value=Stranger(v=1), input_type=Stranger]
```

The error message suggests that Pydantic accepts either a dict or an actual instance of the nested model. Can Ty expand the nested model to something like `Child | Mapping[str, Any]` or `Child | dict` instead of `Any`? Otherwise, writing `Parent(c1=Stranger(v=1))` would not be flagged as an error by Ty, but it will always fail at runtime.

AFAIK, the only case where a different model instance is allowed is when the nested model enables `from_attributes`. In that case, Ty could expand the nested model to `Any` (I believe Ty already parses `model_config`).

> You can already opt out on a per-model (model_config = ConfigDict(strict=True), or class Person(BaseModel, strict=True) or even per-field basis (address: Address = Field(strict=True)). Does that work for your cases? Obviously, that also changes runtime behavior, but it wasn't clear to me if there's a strong use case for "strict checking at type-check time but lax mode behavior at runtime".

I asked for a Ty configuration option because I prefer the type checker to produce false positives at type-check time rather than overlook obvious errors that could cause issues at runtime. Still, narrowing the nested model types would be much better if possible :)

Thank you for taking the time to address this issue!

**Comment by sharkdp** [comment]
@rodrigoscc Thank you for your pushback. I've attempted to improve ty's behavior here. It's not that easy, since there are more exceptions where we need to map to something else than `Child | Mapping[str, Any]`, but I think it can work. If you're interested, it would be great if you could look at the test changes (`pydantic.md`) in https://github.com/astral-sh/ruff/pull/27091 to see if that matches your understanding.

**Comment by rodrigoscc** [comment]
I see there are definitely more exceptions than I thought, haha. Still, that's a great improvement. Thank you!
